File: apps/web/src/lib/predictionConfluence.ts
/**
 * AstroOS — Priority 8: Unified Multi-System Prediction Synthesis & Confluence Client
 *
 * Provides typed interfaces and API communication for:
 * 1. Multi-System Confluence evaluation (k/N agreement, veto inspection)
 * 2. 3-Tier Evidence Provenance (Calculated, Classical, Empirical)
 * 3. Timing Window Intersection (Dasha + Transit + SBC)
 * 4. 1-Click Freeze to P7 Validation Registry
 */

const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://127.0.0.1:8000/api/v1";

export type ProvenanceType =
  | "CALCULATED_EPHEMERIS"
  | "CLASSICAL_LITERATURE"
  | "EMPIRICAL_BACKTEST";

export type SystemSupportStatus =
  | "SUPPORTING"
  | "CONTRADICTING_VETO"
  | "NEUTRAL"
  | "UNAVAILABLE";

export type SynthesizedVerdict =
  | "UNANIMOUS_CONFLUENCE"
  | "STRONG_CONFLUENCE"
  | "MODERATE_CONFLUENCE"
  | "CONFLICTED_VETO"
  | "WEAK_UNCONVERGED";

export type PredictionCategory =
  | "career"
  | "marriage"
  | "finance"
  | "health"
  | "relocation"
  | "education"
  | "spiritual"
  | "general";

export interface SystemContribution {
  system_id: string;
  system_name: string;
  support_status: SystemSupportStatus;
  provenance_type: ProvenanceType;
  primary_houses: number[];
  active_significators: string[];
  rule_or_factor: string;
  rationale: string;
  veto_reason?: string | null;
  evidence_snapshot: Record<string, unknown>;
}

export interface ConfluenceMatrix {
  supporting_count: number;
  veto_count: number;
  neutral_count: number;
  total_systems: number;
  confluence_ratio: number;
  active_vetoes: string[];
  synthesized_verdict: SynthesizedVerdict;
  verdict_rationale: string;
}

export interface SynthesizedTimingWindow {
  window_start: string;
  window_end: string;
  peak_fructification_date: string;
  dasha_sub_period: string;
  transit_trigger: string;
  sbc_trigger_moment: string;
}

export interface EmpiricalTrackRecord {
  historical_hit_rate: number;
  historical_precision?: number | null;
  sample_size: number;
  wilson_95_ci: [number, number];
  sample_size_warning?: string | null;
  matched_cohort_name: string;
}

export interface UnifiedPredictionSynthesis {
  synthesis_id: string;
  chart_id: string;
  subject_name: string;
  category: PredictionCategory;
  synthesized_event_description: string;
  confluence_matrix: ConfluenceMatrix;
  system_contributions: SystemContribution[];
  synthesized_timing_window: SynthesizedTimingWindow;
  empirical_track_record: EmpiricalTrackRecord;
  provenance_breakdown: Record<string, string[]>;
  synthesis_timestamp: string;
  synthesis_hash: string;
}

export interface ConfluenceSynthesisRequest {
  chart_id?: string;
  chart_data?: Record<string, unknown>;
  category: PredictionCategory;
  target_datetime?: string;
  horizon_months?: number;
}

export interface ConfluenceSynthesisResponse {
  synthesis: UnifiedPredictionSynthesis;
}

export interface DomainScanItem {
  category: PredictionCategory;
  event_description: string;
  confluence_verdict: SynthesizedVerdict;
  confluence_ratio: number;
  supporting_count: number;
  veto_count: number;
  active_vetoes: string[];
  peak_timing: string;
}

export interface ConfluenceDomainScanResponse {
  chart_id: string;
  subject_name: string;
  scanned_domains: DomainScanItem[];
  scan_timestamp: string;
}

export interface FreezeToP7Request {
  synthesis_id: string;
  synthesis_payload?: UnifiedPredictionSynthesis;
  target_split_type?: "RESEARCH_TRAIN" | "VALIDATION" | "TEST_OUT_OF_SAMPLE";
}

export interface FreezeToP7Response {
  prediction_id: string;
  chart_id: string;
  subject_name: string;
  technique: string;
  category: PredictionCategory;
  evidence_hash: string;
  frozen_timestamp: string;
  status: string;
  message: string;
}

import { api } from "@/lib/api";

export async function synthesizePrediction(
  req: ConfluenceSynthesisRequest
): Promise<ConfluenceSynthesisResponse> {
  return api.post<ConfluenceSynthesisResponse>("/api/v1/predictions/confluence/synthesize", req);
}

export async function scanConfluenceDomains(
  req: { chart_id?: string; chart_data?: Record<string, unknown>; horizon_months?: number }
): Promise<ConfluenceDomainScanResponse> {
  return api.post<ConfluenceDomainScanResponse>("/api/v1/predictions/confluence/scan", req);
}

export async function freezeToP7Validation(
  req: FreezeToP7Request
): Promise<FreezeToP7Response> {
  return api.post<FreezeToP7Response>("/api/v1/predictions/confluence/freeze-to-p7", req);
}

File: apps/api/schemas/prediction_confluence.py
"""
AstroOS — Prediction Confluence & Synthesis Schemas (Module 23, Priority 8)

Pydantic schemas for multi-system prediction synthesis, cross-domain scanning,
and freezing into immutable P7 validation snapshots.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional
from pydantic import BaseModel, Field

from apps.api.domain.prediction_confluence import (
    ProvenanceType,
    SynthesizedVerdict,
    SystemSupportStatus,
)
from apps.api.domain.prediction_validation import (
    PredictionCategory,
    TemporalSplitType,
)


class SystemContributionSchema(BaseModel):
    system_id: str
    system_name: str
    support_status: SystemSupportStatus
    provenance_type: ProvenanceType
    primary_houses: list[int]
    active_significators: list[str]
    rule_or_factor: str
    rationale: str
    veto_reason: Optional[str] = None
    evidence_snapshot: dict[str, Any] = Field(default_factory=dict)


class ConfluenceMatrixSchema(BaseModel):
    supporting_count: int
    veto_count: int
    neutral_count: int
    total_systems: int
    confluence_ratio: float
    active_vetoes: list[str]
    synthesized_verdict: SynthesizedVerdict
    verdict_rationale: str


class SynthesizedTimingWindowSchema(BaseModel):
    window_start: datetime
    window_end: datetime
    peak_fructification_date: datetime
    dasha_sub_period: str
    transit_trigger: str
    sbc_trigger_moment: str


class EmpiricalTrackRecordSchema(BaseModel):
    historical_hit_rate: float
    historical_precision: Optional[float] = None
    sample_size: int
    wilson_95_ci: tuple[float, float]
    sample_size_warning: Optional[str] = None
    matched_cohort_name: str = ""


class UnifiedPredictionSynthesisSchema(BaseModel):
    synthesis_id: str
    chart_id: str
    subject_name: str
    category: PredictionCategory
    synthesized_event_description: str
    confluence_matrix: ConfluenceMatrixSchema
    system_contributions: list[SystemContributionSchema]
    synthesized_timing_window: SynthesizedTimingWindowSchema
    empirical_track_record: EmpiricalTrackRecordSchema
    provenance_breakdown: dict[str, list[str]]
    synthesis_timestamp: datetime
    synthesis_hash: str


class ConfluenceSynthesisRequest(BaseModel):
    chart_id: Optional[str] = None
    chart_data: Optional[dict[str, Any]] = None
    category: PredictionCategory = PredictionCategory.CAREER
    target_datetime: Optional[datetime] = None
    horizon_months: int = Field(default=12, ge=1, le=60)


class ConfluenceSynthesisResponse(BaseModel):
    synthesis: UnifiedPredictionSynthesisSchema


class ConfluenceDomainScanRequest(BaseModel):
    chart_id: Optional[str] = None
    chart_data: Optional[dict[str, Any]] = None
    target_datetime: Optional[datetime] = None
    horizon_months: int = Field(default=12, ge=1, le=60)


class DomainScanItem(BaseModel):
    category: PredictionCategory
    event_description: str
    confluence_verdict: SynthesizedVerdict
    confluence_ratio: float
    supporting_count: int
    veto_count: int
    active_vetoes: list[str]
    peak_timing: datetime


class ConfluenceDomainScanResponse(BaseModel):
    chart_id: str
    subject_name: str
    scanned_domains: list[DomainScanItem]
    scan_timestamp: datetime


class FreezeToP7Request(BaseModel):
    synthesis_id: str
    synthesis_payload: Optional[UnifiedPredictionSynthesisSchema] = None
    target_split_type: TemporalSplitType = TemporalSplitType.VALIDATION


class FreezeToP7Response(BaseModel):
    prediction_id: str
    chart_id: str
    subject_name: str
    technique: str
    category: PredictionCategory
    evidence_hash: str
    frozen_timestamp: datetime
    status: str
    message: str
